lib/Doctrine/ORM/Persisters/OneToManyPersister.php

```diff
@@ -157,50 +162,19 @@ public function slice(PersistentCollection $coll, $offset, $length = null)
      */
     public function containsKey(PersistentCollection $coll, $key)
     {
-        list($quotedJoinTable, $whereClauses, $params) = $this->getJoinTableRestrictions($coll, true);
-
-        $mapping     = $coll->getMapping();
-        $sourceClass = $this->em->getClassMetadata($mapping['sourceEntity']);
-
-        $whereClauses[] = $sourceClass->getColumnName($mapping['indexBy']) . ' = ?';
-        $params[] = $key;
-
-        $sql = 'SELECT 1 FROM ' . $quotedJoinTable . ' WHERE ' . implode(' AND ', $whereClauses);
-
-        return (bool) $this->conn->fetchColumn($sql, $params);
-    }
-
-    private function getJoinTableRestrictions(PersistentCollection $coll, $addFilters)
-    {
-        $mapping     = $coll->getMapping();
-        $targetClass = $this->em->getClassMetadata($mapping['targetEntity']);
-        $sourceClass = $this->em->getClassMetadata($mapping['sourceEntity']);
-        $id          = $this->em->getUnitOfWork()->getEntityIdentifier($coll->getOwner());
-
-        $whereClauses = array();
-        $params       = array();
-
-        $joinColumns = $targetClass->associationMappings[$mapping['mappedBy']]['joinColumns'];
-        foreach ($joinColumns as $joinColumn) {
-            $whereClauses[] = $joinColumn['name'] . ' = ?';
-
-            $params[] = ($targetClass->containsForeignIdentifier)
-                ? $id[$sourceClass->getFieldForColumn($joinColumn['referencedColumnName'])]
-                : $id[$sourceClass->fieldNames[$joinColumn['referencedColumnName']]];
-        }
+        $mapping   = $coll->getMapping();
+        $uow       = $this->em->getUnitOfWork();
+        $persister = $uow->getEntityPersister($mapping['targetEntity']);
 
-        if ($addFilters) {
-            $filterTargetClass = $this->em->getClassMetadata($targetClass->rootEntityName);
-            foreach ($this->em->getFilters()->getEnabledFilters() as $filter) {
-                if ($filterExpr = $filter->addFilterConstraint($filterTargetClass, 't')) {
-                    $whereClauses[] = '(' . $filterExpr . ')';
-                }
-            }
-        }
+        // only works with single id identifier entities. Will throw an
+        // exception in Entity Persisters if that is not the case for the
+        // 'mappedBy' field.
+        $criteria = new Criteria();
 
-        $quotedJoinTable = $this->quoteStrategy->getTableName($targetClass, $this->platform) . ' t';
+        $criteria->andWhere(Criteria::expr()->eq($mapping['mappedBy'], $coll->getOwner()));
```

Comment:
Since we are in internals, we could avoid using the expression builder, and just passing in the built expression